lisa/transformers/kernel_installer.py

```diff
@@ -0,0 +1,308 @@
+        # 4.18.0-1025.27~18.04.1 amd64 [installed]
+        # output 4.18.0-1025
+        kernel_version_package_pattern = re.compile(
+            f"^{source}/[^ ]+ "
+            r"(?P<kernel_version>[^.]+\.[^.]+\.[^.-]+[.-][^.]+)\..*[\r\n]+",
+            re.M,
+        )
+        # the grep is used to remove color code
+        result = node.execute(f"apt search {source} | grep ''", shell=True)
+        kernel_version = get_matched_str(result.stdout, kernel_version_package_pattern)
+        assert kernel_version, (
+            f"cannot find kernel version from apt results by pattern: "
+            f"{kernel_version_package_pattern.pattern}"
+        )
+        self._log.info(f"installed kernel version: {kernel_version}")
+
+        return kernel_version
+
+    def _replace_boot_entry(self, kernel_version: str, node: RemoteNode) -> None:
+        self._log.info("updating boot menu...")
+        ubuntu: Ubuntu = cast(Ubuntu, node.os)
+
+        # set installed kernel to default
+        #
+        # get boot entry id
+        # postive example:
+        #         menuentry 'Ubuntu, with Linux 5.11.0-1011-azure' --class ubuntu
+        # --class gnu-linux --class gnu --class os $menuentry_id_option
+        # 'gnulinux-5.11.0-1011-azure-advanced-3fdd2548-1430-450b-b16d-9191404598fb' {
+        #
+        # negative example:
+        #         menuentry 'Ubuntu, with Linux 5.11.0-1011-azure (recovery mode)'
+        # --class ubuntu --class gnu-linux --class gnu --class os $menuentry_id_option
+        # 'gnulinux-5.11.0-1011-azure-recovery-3fdd2548-1430-450b-b16d-9191404598fb' {
+        cat = node.tools[Cat]
+        menu_id_pattern = re.compile(
+            r"^.*?menuentry '.*?(?:"
+            + kernel_version
+            + r"[^ ]*?)(?<! \(recovery mode\))' "
+            r".*?\$menuentry_id_option .*?'(?P<menu_id>.*)'.*$",
+            re.M,
+        )
+        result = cat.run("/boot/grub/grub.cfg")
+        submenu_id = get_matched_str(result.stdout, menu_id_pattern)
+        assert submenu_id, (
+            f"cannot find sub menu id from grub config by pattern: "
+            f"{menu_id_pattern.pattern}"
+        )
+        self._log.debug(f"matched submenu_id: {submenu_id}")
+
+        # get first level menu id in boot menu
+        # input is the sub menu id like:
+        # gnulinux-5.11.0-1011-azure-advanced-3fdd2548-1430-450b-b16d-9191404598fb
+        # output is,
+        # gnulinux-advanced-3fdd2548-1430-450b-b16d-9191404598fb
+        menu_id = self.__menu_id_parts_pattern.sub(
+            r"\g<prefix>-\g<postfix>", submenu_id
+        )
+        assert menu_id, f"cannot composite menu id from {submenu_id}"
+
+        # composite boot menu in grub
+        menu_entry = f"{menu_id}>{submenu_id}"
+        self._log.debug(f"composited menu_entry: {menu_entry}")
+
+        ubuntu.set_boot_entry(menu_entry)
+        node.execute("update-grub", sudo=True)
```

Comment:
I still wonder how are those distros setup WRT sudo: *every one of them* with something like ``%wheel        ALL=(ALL)       NOPASSWD: ALL`` on /etc/sudoers? At least Fedora is not like that by default. Do people change, when ingressing images on Azure?

Reply:
The Azure images uses cloud-init or wala-agent to initialize. Users can specify to use password or private key. In LISA, it uses private key by default, and there is no password needed.